packages/flutter/lib/src/widgets/sliver.dart

```diff
@@ -544,22 +608,67 @@ class SliverChildListDelegate extends SliverChildDelegate {
   /// The widgets to display.
   final List<Widget> children;
 
+  final Iterator<Widget> _childIterator;
+
+  final Map<Key, int> _mapKeyToIndex;
+
+  final bool _isConstantInstance;
+
+  int _findItemIndex(Key key){
+    if(_isConstantInstance) {
+      return -1;
+    }
+    // This function should only be called when any child has associated key.
+    // We do lazy loading in this function.
+    if(!_mapKeyToIndex.containsKey(key)) {
+      int index;
+      while(_childIterator.moveNext()) {
+        final Widget child = _childIterator.current;
+        index ??= children.indexOf(child);
+        if (child.key != null) {
+          _mapKeyToIndex[child.key] = index;
+        }
+        if (child.key == key) {
+          break;
+        }
+        index++;
+      }
+    }
+    return _mapKeyToIndex.containsKey(key) ? _mapKeyToIndex[key] : -1;
+  }
+
+  @override
+  int findIndexByKey(Key key) {
+    assert(key != null);
+    Key childKey;
+    if (addAutomaticKeepAlives || addRepaintBoundaries || addSemanticIndexes) {
```

Comment:
This may be error-prone. If we add more "add*" calls, we may forget to update this here. Could you potentially create a private subclass of ValueKey (e.g. "class _SaltedValueKey extends ValueKey { }") and use it in the build method? In this method you could then just check "is my key a _SaltedValueKey, if so use its value if not use the Key directly"?